File: src/ru/edu/vtisov/stack/NumberOfValidSubarrays.java
package ru.edu.vtisov.stack;

import java.util.Stack;

// https://leetcode.com/problems/number-of-valid-subarrays

public class NumberOfValidSubarrays {
    public int validSubarrays(int[] nums) {
        int answer = 0;
        Stack<Integer> stack = new Stack<>();
        for (int i = 0; i < nums.length; i++) {
            while(!stack.isEmpty() && nums[i] < nums[stack.peek()]) {
                answer += i - stack.pop();
            }
            stack.push(i);
        }

        while(!stack.isEmpty()) {
            answer += nums.length - stack.pop();
        }
        return answer;
    }
}

File: src/ru/edu/vtisov/stack/monotonic-stack/number-of-valid-subarrays.go
package monotonic_stack

// https://leetcode.com/problems/number-of-valid-subarrays/

func validSubarrays(nums []int) int {
	stack := make([]int, 0)
	result := 0

	for i, n := range nums {
		for len(stack) > 0 && nums[stack[len(stack)-1]] > n {
			result += i - stack[len(stack)-1]
			stack = stack[:len(stack)-1]
		}
		stack = append(stack, i)
	}

	for _, n := range stack {
		result += len(nums) - n
	}
	return result
}
